pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java

```diff
@@ -669,18 +675,105 @@ public void close() throws PulsarClientException {
     @Override
     public void shutdown() throws PulsarClientException {
         try {
-            lookup.close();
-            cnxPool.close();
-            timer.stop();
-            externalExecutorProvider.shutdownNow();
-            internalExecutorService.shutdownNow();
-            conf.getAuthentication().close();
+            // We will throw the last thrown exception only, though logging all of them.
+            Throwable throwable = null;
+            if (lookup != null) {
+                try {
+                    lookup.close();
+                } catch (Throwable t) {
+                    log.warn("Failed to shutdown lookup", t);
+                    throwable = t;
+                }
+            }
+            try {
+                // Shutting down eventLoopGroup separately because in some cases, cnxPool might be using different
+                // eventLoopGroup.
+                shutdownEventLoopGroup(eventLoopGroup);
```

Comment:
If connectionPool and PulsarClient is using the same eventLoopGroup, this might cause call shutDown on it multiple times, not sure what will happen in this situation? Probably move it to after closing `cnxPool`?

Reply:
shouldn't cause any issue as I am checking for already shutdown case. If so, I will not attempt to shut it down. Though I can move it below if you suggest.